Soothsayer sacrifice fix / Refactor WorldInput Use

the-m4a opened the pull request:
I mentioned in the issue thread (#28 ) that there were more than a few differences between the in-game version of WorldInput.lua and the one in CQUI that was replacing it.  I compared the two files and noted all the differences I could find, and worked to whittle down to this model, which I think is more maintainable.

I have tested this with Vanilla, RF, and GS, but considering the scope of the changes I would like additional testing before we complete this push to master.

**WHAT THIS DOES**
- Can now properly do the sacrifice with the Soothsayer
- WorldInput is now hooked instead of replaced entirely (more managable)
- Launchbar is now hooked instead of replaced (more managable)

**WHAT CHANGED, DETAILS**
- Added CQUI WorldInput wrapper lua scripts, for Vanilla, Exp1, Exp2.
  - Separated out the functions that are extensions of the base game versions, the ones that replace them completely, and the unique to CQUI
  - Which means the unmodified game lua is what is used for the Sacrifice stuff
- Removed the WorldInput.lua replacement file
- Did the same for LaunchBar.lua (wrappers for Vanilla, Exp1, Exp2...)
- Removed the LaunchBar.lua replacement file
- Removed some of the unnecessary Key Mappings
- Added "ALT+E" mapping for telling a unit to go explore
- Cleaned up the hotkey code (now using the Hash variable, instead of the actual Hash value)

**HOW TESTED**
- Started various flavor games (Vanilla, RF, GS)
- Regular World View: Verified map dragging worked as expected
- City Screen: if starting a map drag while mouse pointer was on top of Tile Purchase icon, that tile was NOT purchased (base game would typically result in the tile being purchased)
  - Side note: I find myself thinking we may want to make this a configurable option, as even the _slightest_ map drag when trying to purchase a tile will prevent the purchase.  Making it optional would be good for accessibility, and I don't know offhand how much of an issue it creates for touch interfaces.
- Regular World View: Verified functional: direct a unit to move to a place more than 1 turn away, allow UI to go automatically to next unit, go back to previous unit and verify the "queued path" (where the unit is going to) shows as expected
- Regular World View: Verified I can sacrifice units with the Soothsayer during the Emergency
- Regular World View: verified WASD and Arrow key functionality for moving about the map
- Hot Keys were tested and all but Air-Strike seem to be ok.  I'm not convinced Air-Strike ever worked.

**WHAT ELSE IS NEEDED**
- Peeps to play this version and make sure it's good

the-m4a commented:
@alimulhuda @alexeyOnGitHub if you could give a looksie, I would appreciate it!

alexeyOnGitHub commented:
the approach described in the PR seems reasonable to me. I prefer to only keep deltas and not the whole clone of the Civ 6 UI, with a few changed lines here and there.
this should lower the cost of  the ongoing mod maintenance.

no strong opinion on "ALT+E" shortcut. I usually use "autoexplore" operation once or twice per game, so I would skip this shortcut and maybe save this keys combination for something else instead.  but as long as it is configurable in some menu, this should not be a problem and  it would  be up to the user whether or not they care about it.

you mentioned that you removed some other hotkey mappings. what are they?

no opinion on the code changes themselves. I can download and run this version of the mod over the weekend.

the-m4a commented:
Removed the ones that already had support in game - not sure if those were introduced by firaxis after the initial CQUi mod was introduced but it didn’t seem to make sense to duplicate the effort

Specifically: arrow keys, ‘z’ (for sleep), default ‘f’ handler (there’s special cases for F, with alt and shift I think and the CQUi code would also handle the F default which wasn’t necessary...)
Q (for quarry)..

In each case I commented the code out first and tested it, in vanilla and GS.  Once I got to this set of changes you see I also tried RF.

There may still be something to investigate - I couldn’t figure why “G” was not getting handled by the game when CQUi keyboard is in “enhanced” mode (the default).  I had added prints and the like to make sure the key was getting passed out of the CQUI key up function  but it didn’t seem to work.  Works when enhanced mode is off which is weird...
I will double check this tonight

Also 2 other things
1 I learned the ScriptReplace tags in the modinfo is case sensitive for the file names, which is silly on a windows machine.  Maybe makes sense for linux?

I will open up a new issue as I want to focus on getting those replacement files out as well.  A lot more maintainable when we are not comparing 4000 line files to try and spot the diff


Sent from my iPhone

On Jun 12, 2020, at 2:45 PM, Alexey <notifications@github.com> wrote:

﻿

the approach described in the PR seems reasonable to me. I prefer to only keep deltas and not the whole clone of the Civ 6 UI, with a few changed lines here and there.
this should lower the cost of the ongoing mod maintenance.

no strong opinion on "ALT+E" shortcut. I usually use "autoexplore" operation once or twice per game, so I would skip this shortcut and maybe save this keys combination for something else instead. but as long as it is configurable in some menu, this should not be a problem and it would be up to the user whether or not they care about it.

you mentioned that you removed some other hotkey mappings. what are they?

no opinion on the code changes themselves. I can download and run this version of the mod over the weekend.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://eur06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fcivfanatics%2FCQUI_Community-Edition%2Fpull%2F35%23issuecomment-643495695&data=02%7C01%7C%7C93f0f4d88b5146e16abd08d80f19ee3e%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637275951325320180&sdata=IEDV6t1f%2FlzhNunhYSoOzhg13CoIHH82X%2BIgDk3Xbyw%3D&reserved=0>, or unsubscribe<https://eur06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACDBNOASKKHSZCS4XW5PPIDRWKOXVANCNFSM4NZUGA7A&data=02%7C01%7C%7C93f0f4d88b5146e16abd08d80f19ee3e%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637275951325320180&sdata=dEi0Ghc9xtFDnMBvg%2FCR7%2BeXNpNF3cr1R%2F7j7Zf%2F2Dg%3D&reserved=0>.

wayneb64 commented:
While we seem to have a lot of help on CQUI I would like to request that someone take a look at the build queue. The CQUI version was removed when the Vanilla game added it's own version but it doesn't work right. It won't let you queue a Harbor AND a Lighthouse and every time you capture a city you have to keep coming back and adding damaged buildings to the queue rather than all at once. Really annoying to me as I take a LOT of cities. 

I know getting the mod back to functional takes priority but this would really make it better in my opinion. Also, I have seen a ton of activity but it's not clear to me what percent of the mod has been repaired. How close to complete is it or what is still not done? I have a test box setup but I am still playing offline with the old mod.

alimulhuda commented on CQUI.modinfo:
```diff
@@ -1,12 +1,12 @@
     <Teaser>CQUI</Teaser>
     <Description>The Community Quick User Interface MOD, CivFanatics Version: Updated May 2020.</Description>
-    <CompatibleVersions>2.0,2.01</CompatibleVersions>
+    <CompatibleVersions>2.0,2.01,2.1</CompatibleVersions>
```
I think this line is for indicating compatibility with different versions of Civilization 6. I tried building one of the sample projects in ModBuddy and the project properties have options for indicating compatibility with the two expansions. The line should be
```suggestion
    <CompatibleVersions>1.2,2.0</CompatibleVersions>
```
1.2 is for Rise and Fall - Spring Update
2.0 is for Gathering Storm

the-m4a replied:
I'll fix this now

alimulhuda commented on CQUI.modinfo:
```diff
@@ -1,12 +1,12 @@
 <?xml version="1.0" encoding="utf-8"?>
-<Mod id="1d44b5e7-753e-405b-af24-5ee634ec8a01" version="2.1">
+<Mod id="1d44b5e7-753e-405b-af24-5ee634ec8a01" version="2.2">
```
ModBuddy GUI only shows integer version numbers. It allows you to input floats but then rounds it and shows whole number. Building the mod shows whatever version you actually typed in the box. Just FYI, something I noticed. Not sure if we should do anything about it though.

the-m4a replied:
I guess just for our use, I suppose?  Yeah if it doesn't do anything I don't know how important it'd be...

alimulhuda commented:
@wayneb64 It should be functional from a while ago (if you ignore the soothsayer issue that the new DLC introduces). I've been playing with what we have in master branch and it seems to be no worse than it was before the May patch. More things are slowly getting fixed. I don't have the new DLC though, only have the two expansions and a few of the civilization packs.
Regarding your request, I think you should file it in the issues section (https://github.com/civfanatics/CQUI_Community-Edition/issues) so we can track it and get it fixed.

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+local CQUI_HOTKEYMODE_CLASSIC     = 1;
+local CQUI_HOTKEYMODE_STANDARD    = 0;
+
+local m_isInputBlocked  :boolean = false;
```
Does CQUI need this variable? Unmodded version removed some usages two patches ago in September patch and completely removed it in the last patch.

the-m4a replied:
not sure - didn't investigate, but if not, then I'm ok to remove it.  I'll look and see...

the-m4a replied:
I think you're right, will remove it.

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+local CQUI_hotkeyMode   :number  = CQUI_HOTKEYMODE_ENHANCED;
+local CQUI_isShiftDown  :boolean = false;
+local CQUI_isAltDown    :boolean = false;
+local CQUI_isCtrlDown   :boolean = false;
```
Doesn't seem to be used anywhere.

the-m4a replied:
yep, added it with intention of using it, then didn't use it after reviewing some more.
removed.

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+-- Each of these functions call the Base Function found in WorldInput.lua
+-- ===========================================================================
+
+-- ===========================================================================
```
Any idea why the original modified this function instead of `StartDragMap`?

the-m4a replied:
not sure.  I don't see a reason why it couldn't be in StartDragMap - all the instances where StartDragMap is called, it's preceded by a check whether the mouse/touch is already dragging, which if so, it calls UpdateDragMap.  

That said, this works, and while logically it may make more sense to move it, I'm not sure it's necessary?

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+end
+
+-- ===========================================================================
+function RealizeMovementPath(showQueuedPath:boolean)
```
Looks like the ability to see unit path on hover over unit is being removed. Is it intentional? If yes need to remove CQUI option that says hover on unit will show unitpath and also clear up related function calls (CQUI_OnUnitFlagPointerEntered in unitflagmanager_CQUI.lua).
My preference is to keep the hover functionality (unless the unmodded game has that functionality now), it comes in handy when planning movements with lots of units.

the-m4a replied:
yeah, I was wondering about that, and had intended to look into it.  Good catch.  I would also prefer the hover show the planned movement.

the-m4a replied:
Actually - a little confused here... are you saying the base game is removing the show path on hover?

alimulhuda replied:
The base game doesn't have the ability to show path on hover. It was added by Azurency by adding the function I mentioned in original comment and by adding a new parameter for this function. Your modification looks like it removes the CQUI ability to show unit path without selecting the unit.

the-m4a replied:
Aaah - okay, it wasn't anything specific in the RealizeMovementPath function... I think it may have been that I was calling the base initialize after Adding methods for the LuaEvents.  Checking into that now.

the-m4a replied:
Argh, ok, I see it.  Was hoping to avoid having a huge function in there, but it seems in order to achieve this, we need to have the huge function in there.

the-m4a replied:
had to put in the entire function, but it works now!

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+    CQUI_isAltDown = true;
+  end
+
+  if CQUI_hotkeyMode ~= CQUI_HOTKEYMODE_STANDARD then
```
Seems unnecessary due to the condition below.

the-m4a replied:
fixed

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+  if (keyPanChanged == true) then
+    -- Base game file uses m_edgePanX and m_edgePanY... but those values do not appear to ever change from 0, so just send 0.
+    ProcessPan(0, 0);
+    return true;
```
Line above seems new. Any reason to add it?

the-m4a replied:
it's relocated from the original end of that function.  The original end of the function also had handlers for the arrow keys, before calling this method.  I removed the arrow key handlers (just let the base game handle them), it didn't make sense to have the ProcessPan call be at the bottom of that function anymore, so i moved it closer to the code elements that use it

alimulhuda replied:
I was referring to `return true;`. Doesn't the original always `return false;`?

alimulhuda replied:
I think this is answered by your answer to my other my comment regarding key stroke being consumed.

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+      keyPanChanged = true;
+    end
+
+    if (uiKey == Keys.S) then
```
Couldn't this overlap with the Alt+S handling near end of this function. I'm thinking it might result in rebase and map move if you press Alt+S

the-m4a replied:
hmm I think you're right - good catch - if holding down alt and letting go of 'S', it may catch this here.  Will add the if altdown==false clause

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+    end
+
+    if (uiKey == Keys.Q) then
+      if (CQUI_isAltDown == true and unitType == "UNIT_BUILDER" ) then
```
I think old version would allow Alt+Q for all hotkey modes except standard. Have to take care it doesn't cause map move + build quarry if you do relocate it.

the-m4a replied:
Good catch - I think I intended to move the Alt-Q handler and forgot to put it back.

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+      cquiHandledKey = true;
+    elseif (unitType == "UNIT_MILITARY_ENGINEER") then
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.Improvements["IMPROVEMENT_FORT"].Hash);
+      cquiHandledKey = true;
```
Missing Fortify unit action. Or is it one of the ones you intentionally removed?
```
    else
      UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), -744032280); --Sleep
```

the-m4a replied:
yeah - if it's anything besides a builder or engineer, it'll just pass through without handling the key

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+    end
+  end
+
+  if (uiKey == Keys.H and unitType == "UNIT_BUILDER") then
```
I think original had the H key assigned to fortify until healed for damaged military/religious units. Was this one of the ones you intentionally removed?
```
  if( uiKey == Keys.H ) then
    if (unitType ~= nil) then -- OPTIMIZATION -- if unit health 100% alert not heal
      UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), 2126026491); -- OH MY GOD IT WAS SO HARD TO FIND THAT ID. WHAT WAS FIRAXIS THINKING
    end
    return;
  end
```

the-m4a replied:
Yep - the comment there was funny though.  The game handles the H key in the same manner, so no reason to handle it here.

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+
+  if (uiKey == Keys.P) then
+    if (CQUI_isShiftDown) then
+      CQUI_isShiftDown = false;
```
The clearing of the shift key flag is from old code, but do you know if it serves a purpose? I was thinking that this is the only reason why the clearing of the shift and alt key flags had to be placed in two places. So I was just curious if you knew why.

the-m4a replied:
not sure, actually.  looking at it, this seems like if you kept shift down it would only respond to the first.

and the more I look at it, I'm not positive I need to do cquiHandledKey=false on any of the keys involving the Alt.  When Alt is the KeyUp, then it'd be handled.

the-m4a replied:
removed it there, seems to work ok.

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+      cquiHandledKey =false;
+    end
+  end
+
```
Missing Sleep unit action. Intentional or unintended?
```
  if( uiKey == Keys.Z ) then
    UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), -41338758); --Sleep
  end
```

the-m4a replied:
intentional, game handles Z as sleep

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+    cquiHandledKey = BASE_CQUI_DefaultKeyUpHandler( uiKey );
+  end
+
+  return cquiHandledKey;
```
Old version would only return true when '( uiKey == Keys.VK_ADD or uiKey == Keys.VK_SUBTRACT )'. Any reason for change? Is this supposed to return true whenever a keystroke is consumed?

the-m4a replied:
yep.  when the keystroke is handled, returning true prevents it from going to other handlers (for example, pushing W in Exp2.  When you have a great work, W is the default show Great Works key.  If we do not return true from this handler, then the Great Work panel is opened.  

Which reminds me, adding Alt-W to open the great work panel may be a thing to do

the-m4a replied:
I've decided against the Alt-W thing for the time being

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+  print_debug("** Function Entry: ClearAllCachedInputState (CQUI Hook)");
+  BASE_CQUI_ClearAllCachedInputState();
+
+  CQUI_isShiftDown = false;
```
Should probably also clear the other modifier key states here.
```suggestion
  CQUI_isAltDown = false;
  CQUI_isShiftDown = false;
```

the-m4a replied:
Yep good catch

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+-- ===========================================================================
+function OnInputActionTriggered( actionId:number )
+ -- CQUI Enhanced Keyboard over-rides some of the ingame hot keys
+
```
Is it safe and intentional to leave this function empty? The original seemed to only modify the part which handled previous city and next city.

the-m4a replied:
This is something I forgot to remove - I was messing with it when trying to figure out the issue with the arrow keys and what-not.  Removed it

the-m4a replied:
and this is totally why the "G" key was not being handled!  aaah!!! thank you for spotting this

the-m4a replied:
yep, left by mistake earlier.  Removed.  This was the issue preventing "G" from showing/hiding the grid!

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,559 @@
+ -- CQUI Enhanced Keyboard over-rides some of the ingame hot keys
+
+end
+
```
Unmodded has a new function around here. We should also probably clear our alt key state flag in the same situation
```
-- Alt-F4 in some situations can cause m_isALTDown to be stuck on after you click "Cancel", so clear it here.
function OnUserRequestClose()
  m_isALTDown  = false;
end
```

the-m4a replied:
good call

alimulhuda commented on Assets/UI/launchbar_CQUI.lua:
```diff
@@ -0,0 +1,250 @@
+end
+
+-- ===========================================================================
+function TestLaunchBarExtension()
```
Looks like a function for experiments. I'm just adding this note as a reminder to remove before merging to master.

the-m4a replied:
it was in the previous set of code - commented out.   I added it as a function, but this function is never called.

This is the LaunchBar thing that we don't use at the moment - I figured can do this change and then remove it later if we want to remove it entirely, so we at least have the history.

the-m4a replied:
added a comment, left the function

alimulhuda commented on Assets/UI/launchbar_CQUI.lua:
```diff
@@ -0,0 +1,250 @@
+  BASE_CQUI_OnOpen();
+end
+
+-- ===========================================================================
+function Unsubscribe()
+  BASE_CQUI_Unsubscribe();
+  -- Vaniila-only function, because OnInputActionTriggered may be overriden elsewhere (like XP1)
+  Events.InputActionTriggered.Remove( OnInputActionTriggeredVanilla );
+end
+
+-- ===========================================================================
+function Subscribe()
+  BASE_CQUI_Subscribe();
+  -- Undo the default for this Event
+  Events.InputActionTriggered.Remove( OnInputActionTriggered );
+  -- Vaniila-only function, because OnInputActionTriggered may be overriden elsewhere (like XP1)
+  Events.InputActionTriggered.Add( OnInputActionTriggeredVanilla );
+end
+
+-- ===========================================================================
+-- CQUI: Unique Functions
+-- These functions are unique to CQUI, do not exist in the Base WorldInput.lua
+-- ===========================================================================
+function OnInputActionTriggeredVanilla( actionId )
+  -- Vanilla-only function, because OnInputActionTriggered may be overriden elsewhere (like XP1)
+  OnInputActionTriggered( actionId )
+end
```
Are you sure we need these functions? I was looking at the diff between the last version by Azurency (Hash: b73a5fd4) and the base game files that it is supposed to match (Hash: 44400580). It looked like the line with `Events.InputActionTriggered` and even the comment above it was from the unmodded version. In the version after the last Azurency modifications (Hash: 8f28fee5), the matching lines were marked as being CQUI customizations.

the-m4a replied:
Not sure that we need them, no... was just copying over the differences between the two files and this was marked as a difference.  Will remove and test.  I plan to play a game in a bit as I haven't actually played a game in a week, just been starting and playing a few turns to test changes :)

the-m4a replied:
removed them.  updated PR with that change.

the-m4a commented:
Updated the PR based on the feedback - @alimulhuda 
- Mouseover showing the queued path works again (had to use the entire RealizeMovementPath function)
- "G" and other non-handled keys do what they're supposed to again
- Removed some stuff that didn't need to be there anymore

alimulhuda commented on Assets/UI/launchbar_CQUI.lua:
```diff
@@ -0,0 +1,223 @@
+include( "GameCapabilities" );
```
Do we need this? The base game LauchBar.lua includes it and the files that include this file should all include the base game file.

the-m4a replied:
I was trying to figure out the right method for doing this.  Basically, it's an order of operations thing - and I guess there's 2 ways to go about it?  I've seen it be done both ways, I don't know which is better, or if it's necessary to include other things.  Here's my thought process on it:

**BaseGame file and Extension File**
-BaseFileFoo_CQUI_basegame.lua
```
include("BaseFileFoo")  -- the basegame file
include("BaseFileFoo_CQUI.lua") -- our extension/replacer file
```
BaseFileFoo_CQUI.lua
```
include("WhateverThingsTheBaseFileHad?")
```
Whatever globals are defined in _BaseFileFoo_ are available in BaseFileFoo_CQUI because the CQUI file loads after.  The locals in _BaseFileFoo_ are not available outside of that file, though.  

**Just the CQUI Extension File**
BaseFileFoo_CQUI.lua
```
include ("BaseFileFoo")
```

I haven't been able to figure out why do the first one instead of the second, but there was at least one scenario where trying only the second style didn't seem to work and I had to do the one with the BaseFileFoo_CQUI_basegame...

---

Not sure if the extension file (in this case launchbar_CQUI) refers to anything in GameCapabilities, so perhaps in this case it's not needed...

alimulhuda replied:
I think the approach for the first way is better. It allows us to have one file `BaseFileFoo_CQUI.lua` that we can use for all version (Base/Exp1/Exp2). This file contains the actual mod changes. So if we need to extend/replace something that the expansion also extends/replaces, then I think we need the first option.

With the second option it becomes more complex when you need to include `BaseFileFoo_Expansion1`, since that file will most likely already include `BaseFileFoo`. Double includes might result in some lines of code being run more than once (unverified guess), like those that exist outside a function. So i think the second option is only viable if the expansions don't do a ReplaceUIScript with the context in question.

Regarding your comment about
`include("WhateverThingsTheBaseFileHad?")`
I don't think it would be needed since the second way doesn't need it I don't see why the first way would. Both include `include("BaseFileFoo")  -- the basegame file` and should have access to the same things. We can also just remove `include("WhateverThingsTheBaseFileHad?")` and see if we get Lua errors for trying to use undefined functions. I'm concerned about double includes (where the `BaseFileFoo` includes the things we try to include in `BaseFileFoo_CQUI`), in my mind the kind of errors that might come up from doing this would be ones that are subtle and hard to debug.

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,768 @@
+include("PopupDialog.lua");
```
Do we need to include this? The unmodded base game file already includes it.

the-m4a replied:
same as above - if our CQUI version refers to things defined in PopupDialog, then I think it'd need it... but again, not sure.

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,768 @@
+      end
+    end
+
+    if (uiKey == Keys.VK_SHIFT and ContextPtr:LookUpControl("/InGame/TechTree"):IsHidden() and ContextPtr:LookUpControl("/InGame/CivicsTree"):IsHidden()) then
+      cquiHandledKey = true;
+      if (CQUI_cityview) then
+        UI.SetInterfaceMode(InterfaceModeTypes.SELECTION);
+        UI.SelectNextReadyUnit();
+      else
+        LuaEvents.CQUI_GoNextCity();
+      end
+    end
```
This is just an observation, no code change required for this pull.
The unmodded game allows pressing Shift+Enter to force end turn. This means you don't have to cycle through all your units and tell them to skip turn. Very helpful for late game large army games. I think having this section here results in the view jumping around when the Shift key is released after the forced end turn. I'll file an issue. Just wanted to point at code ask what your thoughts were regarding the behaviour.

the-m4a replied:
I am not a fan of the shift-to-show-city thing, but I also don't use the keyboard shortcuts myself.  I know the shift key was not showing the city with the Shift-P (with last night's changes) but didn't try the Shift-Enter

alimulhuda replied:
In general I think modifier keys (Alt, Shift, Ctrl) should be just that, modifiers. They should not be used for any specific action by themselves. Anyways, I'm just biased against this hotkey. I've seen weird behaviour when doing Shift+Enter while running CQUI. I'll try to right up some stuff for why I think this one is a bad idea and under what conditions I see weird behaviour when I file the issue later.

the-m4a replied:
I agree - I'll take it out and add it to that issue for the airbases and stuff, can figure it out later

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,768 @@
+      keyPanChanged = true;
+    end
+
+    if (uiKey == Keys.S and CQUI_isAltDown == false) then
```
I've tried it with this change and I think the old code was better. With the current code you can get a situation where you press Alt, then press S (key down handler starts moving it down now), then release S (the new code will not clear the flag `m_isDOWNpressed` since Alt is still held down), then release Alt. The game will just keep panning down even with no keys held, I was thinking it should stop moving down when Alt is released but for some reason it doesn't.

I think a better change would be to modify the air strike command to something else. Or just comment out or remove since it doesn't seem to work anyway.

the-m4a replied:
I think you're right.. you mentioned below - the strike and the nuke commands require that we're pointing at a unit or tile and by itself... which I guess you could do if the mouse is pointing at it?
But even then... maybe just comment it out with a TODO, figure it out kind of thing

alimulhuda replied:
Yup, makes sense to comment out the air commands. Just don't forget to change this back to 
```suggestion
    if (uiKey == Keys.S) then
```
Otherwise we'll have a bug if someone tries to press Alt+S

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,768 @@
+      -- Build Road/Rail is a UnitOperation
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.UnitOperations["UNITOPERATION_BUILD_ROUTE"].Hash);
+      cquiHandledKey = true;
+    elseif (formationClass == "FORMATION_CLASS_AIR") then
+      UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), UnitOperationTypes.REBASE);
+      cquiHandledKey = true;
+    end
+  end
+
+  if (uiKey == Keys.S and CQUI_isAltDown == true) then
+    if (formationClass == "FORMATION_CLASS_AIR") then
+      UnitManager.RequestCommand(UI.GetHeadSelectedUnit(), UnitOperationTypes.AIR_ATTACK);
+      cquiHandledKey = true;
+    end
```
Can't seem to get either of these air unit commands to work. Did it work for you? Also theoretically the S key one should be `RequestOperation` function right? I tried it with that and it also doesn't work.

alimulhuda replied:
I was looking through the unmodded files for anything related to air unit rebasing. Found these functions `OnMouseRebaseEnd`, `AirUnitReBase` and `OnInterfaceModeChange_ReBase` in the WorldInput.lua file. Also the `RequestOperation` call looks like this
`UnitManager.RequestOperation( pSelectedUnit, UnitOperationTypes.REBASE, tParameters);`.

The stuff related to air attack also seems to be a multi-stage, multi-function process. `UnitAirAttack`, `OnInterfaceModeChange_Air_Attack`.

The WMD stuff also looks like it is multi-stage process, but I haven't tried it yet. Did you try the hotkey? Based on what I'm seeing in the base game code, it shouldn't work since it also requires a third parameter which we don't provide.

Right now I'm thinking it might not be a simple single command that we can issue, if that is the case maybe we should just remove it from this pull and work on it separately.

the-m4a replied:
I opened up Issue #40 to track this.  I'll add your note there, as well.  Commented out the code in the KeyUp handler in the meantime, with a note.

At this point I want to get the fix for the Soothsayer done and think it's an okay tradeoff to lose these hotkeys that were not likely working as intended anyway

alimulhuda commented:
For all the new and modified files, could you end with an empty blank line. GitHub diff complains "No newline at end of file", and in the past I've had some issues with computer lab assignments because some tools assume the file end to be an empty blank line.
Feel free to ignore the this comment, it's just old issues I've had in the past and become a little paranoid sometimes.

the-m4a commented:
Played some last night, this seems to work.  If the PR looks good to you @alimulhuda @alexeyOnGitHub please approve it
If you do the merge, please do squash merge since there's a ton of commits with this one

alimulhuda commented on Assets/UI/worldinput_CQUI.lua:
```diff
@@ -0,0 +1,780 @@
+      -- Build Road/Rail is a UnitOperation
+      CQUI_BuildImprovement(UI.GetHeadSelectedUnit(), GameInfo.UnitOperations["UNITOPERATION_BUILD_ROUTE"].Hash);
+      cquiHandledKey = true;
+    elseif (formationClass == "FORMATION_CLASS_AIR") then
+      UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), UnitOperationTypes.REBASE);
+      cquiHandledKey = true;
```
These 3 lines with the rebase operation should also probably be commented out and handled with Issue #40

the-m4a commented:
Thanks.  After work today I'll do the upload to the Workshop.

alexeyOnGitHub commented:
what would be the best way to test a change like this if you already have some version of the mod installed through Steam? we should document this process in `CONTRIBUTING.md` file. 
e.g. "unsubscribe from the plugin in Steam (or maybe disable in the UI only?), delete its folder in this location on disk (?), clone this repo, place it into a certain folder, hook it up to the game somehow, test.........., then delete the folder, then re-subscribe to the officially released version (on Steam) later". or something like this.

alimulhuda commented:
@alexeyOnGitHub What I do is just have a repo cloned to a folder in \Documents\My Games\Sid Meier's Civilization VI\Mods. Then I switch to the branch I want to test. I don't have the version from workshop subscribed normally.
When I want to test the version from workshop I change the extension of the .modinfo file in my other location, so that the game doesn't recognize it as mod then I subscribe to the workshop one.

If your preferred version is the one from workshop then you could try one of the following.
- Unsubscribe from workshop version then checkout the GitHub version to the My Games mod location.
- Modify the extension for the .modinfo file in the steam workshop location (One of the folders in Steam\steamapps\workshop\content\289070 should have it). Then checkout the GitHub version into the My Games mod location.

the-m4a commented:
Gah, I clicked on the re-review button, thinking it was a refresh button.  @alimulhuda can you click the approve button again?

Also - that's what I do, I cloned the repo to my  \Documents\My Games\Sid Meier's Civilization VI\Mods folder.  I am subscribed as well, though mostly to ensure it updates when I push the update.  I then manually go into the subscribed version and change the GUID (just the first few characters) to AAA, so I differentiate while looking at the mod list.

the-m4a commented:
also wanted to say thanks for the thorough review!  Very helpful, good stuff

alexeyOnGitHub commented:
here is what I did - 
* moved the mod git repo into  \Documents\My Games\Sid Meier's Civilization VI\Mods folder
* edited the plugin name and ID in its ".modinfo" file to make sure the local copy is different from Civ6 point of view.
* started Civ6
* opened "mods" dialog in the game launcher and created a new "mods group".
* unselected the workshop-provided CQUI 
* selected the local version of this mod
* started a new game in Renaissance era, played ~20 turns. 
* I was able to settle a city, choose a pantheon, send a trade route, send envoys, ..
* everything was good until I met some other civilization, then the UI disappeared.

I had the same problem in the past when had a local copy of the plugin that was not compatible with the game anymore, and I wanted to test the workshop-provided one. even though I **explicitly disabled the local mod version** in the game settings, and the workshop-provided copy was supposed to be OK/Fixed for the "new frontier", the UI was disappearing on civ meet. which suggests that a copy of this plugin, even when DISABLED, will still interact or conflict with this plugin provided by the Workshop.

I wonder if this may be a problem with the **plugin loading mechanism**, or there is something we can fix in the **plugin source code**?

alimulhuda commented:
Could you place your Lua log file here it should be in Documents\My Games\Sid Meier's Civilization VI\Logs. Whenever I've had things like this happen the log file gave some clue as to what was happening. I usually search the Log folder for the word "error" and look over the lines where it shows up.

Also I don't trust systems to always work right, that's why I do the thing with extension modification. That way the mod system just doesn't know how to to load the files. It won't allow for your things with the mod groups, but I think it should be more resilient.

alexeyOnGitHub commented:
I uploaded lua.log file here: https://docs.google.com/document/d/1ebfHT2F28CExkWgOu9or1H-UG1PclLdeMdJLC2Rl6yU/edit?usp=sharing
it is pretty big (600+ kb) and I am not sure what part of it is relevant (is it overwritten on each game start?) , so I just uploaded the whole thing.

alimulhuda commented:
I'll have to take a closer look at your log file later. Did a quick look over and it seems like you have at least one other mod (Extended Diplomacy Ribbon), and for some reason it's interfering. It also looks like your version of the mod in My Games is an old one. The error with the EDR mod is one I fixed. But your log file looks like its the version from before my fix.

Are you sure you are using the latest version from GitHub for your tests? Does your modinfo file say it's version 2.2.

the-m4a commented:
Looking now as well, just saw this and happen to be at the machine.

I did update the workshop with the current version, so if there's something here I'll roll it back

the-m4a commented:
If you upload the Modding.log file that also has some useful info - I think that gives the version number and whatnot.

alimulhuda commented:
I did a folder compare and it looks like the workshop version is identical to the version in this branch. So if @alexeyOnGitHub just uses the workshop version it should be the same as the GitHub version.

The reason why I think his MyGames version is outdated is because of the error in the log file 
`Runtime Error: C:\Users\ask\Documents\My Games\Sid Meier's Civilization VI\Mods\CQUI_Community-Edition\Assets\Expansion1\Replacements\LeaderIcon.lua:173: attempt to index a nil value
`
Line 173 in the updated version is an else statement. In the old version it used to be a call to `self.Controls.CQUI_Era:SetText`

the-m4a commented:
Error shows issue happening at line 173 in expansion1\leadericon.lua:
```
	C:\Users\ask\Documents\My Games\Sid Meier's Civilization VI\Mods\CQUI_Community-Edition\Assets\Expansion1\Replacements\LeaderIcon.lua:173: in function 'LeaderIcon:UpdateTeamAndRelationship'
	C:\Users\ask\Documents\My Games\Sid Meier's Civilization VI\Mods\CQUI_Community-Edition\Assets\Expansion1\Replacements\LeaderIcon.lua:69: in function 'LeaderIcon:UpdateIcon'
	C:\Program Files (x86)\Steam\steamapps\common\Sid Meier's Civilization VI\Base\Assets\UI\DiplomacyRibbon.lua:135: in function 'AddLeader'
	C:\Program Files (x86)\Steam\steamapps\workshop\content\289070\1360462633\ExtendedDiplomacyRibbon.lua:467: in function 'AddLeader'
	C:\Program Files (x86)\Steam\steamapps\common\Sid Meier's Civilization VI\Base\Assets\UI\DiplomacyRibbon.lua:216: in function 'UpdateLeaders'
	C:\Program Files (x86)\Steam\steamapps\workshop\content\289070\1360462633\XP2\EDR_DiplomacyRibbon_Expansion2.lua:40: in function 'UpdateLeaders'
	C:\Program Files (x86)\Steam\steamapps\common\Sid Meier's Civilization VI\Base\Assets\UI\DiplomacyRibbon.lua:455: in function 'OnDiplomacyMeet'
```

Before alimulhuda's change, this was line 173:

```
self.Controls.CQUI_Era:SetText("[ICON_GLORY_NORMAL_AGE]");
```

The change made was to check for nil, which wouild make getting to that line now impossible, so that must be an older copy that you are running
```
if (self.Controls.CQUI_Era ~= nil) then
       local gameEras:table = Game.GetEras();
         if gameEras:HasHeroicGoldenAge(playerID) then
           self.Controls.CQUI_Era:SetText("[ICON_GLORY_SUPER_GOLDEN_AGE]");
         elseif gameEras:HasGoldenAge(playerID) then
           self.Controls.CQUI_Era:SetText("[ICON_GLORY_GOLDEN_AGE]");
         elseif gameEras:HasDarkAge(playerID) then
           self.Controls.CQUI_Era:SetText("[ICON_GLORY_DARK_AGE]");
        else
          self.Controls.CQUI_Era:SetText("[ICON_GLORY_NORMAL_AGE]");
      end
  end
````

alexeyOnGitHub commented:
terribly sorry for the confusion, guys! I launched this on a computer that had an old clone of the repo. the problem is gone after I pulled the latest changes from this branch.
thanks a lot for the quick response!

the-m4a commented:
no worries!  I'll squash merge this now :)
